core/src/Streamly/Internal/Unicode/Stream.hs

```diff
@@ -682,14 +745,100 @@ decodeUtf8Lax = decodeUtf8
 -- Decoding Utf16
 -------------------------------------------------------------------------------
 
+data UTF16CharDecodeState
+    = UTF16CharDecodeInit
+    | UTF16CharDecoding !Word16
+
+{-# INLINE parseCharUtf16With #-}
+parseCharUtf16With ::
+       Monad m => CodingFailureMode -> ParserD.Parser Word16 m Char
+parseCharUtf16With cfm = ParserD.Parser step initial extract
+
+    where
+
+    prefix = "Streamly.Internal.Unicode.Stream.parseCharUtf16With:"
+
+    {-# INLINE initial #-}
+    initial = return $ ParserD.IPartial UTF16CharDecodeInit
+
+    handleError err souldBackTrack =
+        case cfm of
+            ErrorOnCodingFailure -> ParserD.Error err
+            TransliterateCodingFailure ->
+                case souldBackTrack of
+                    True -> ParserD.Done 1 replacementChar
+                    False -> ParserD.Done 0 replacementChar
+            DropOnCodingFailure ->
+                case souldBackTrack of
+                    True -> ParserD.Continue 1 UTF16CharDecodeInit
+                    False -> ParserD.Continue 0 UTF16CharDecodeInit
+
+    {-# INLINE combineSurrogates #-}
+    combineSurrogates hi lo =
+        let first10 = fromIntegral (hi - utf16HighSurrogate) `shiftL` 10
+            second10 = fromIntegral (lo - utf16LowSurrogate)
+         in unsafeChr (0x10000 + (first10 .|. second10))
+
+    -- In little endian we should get low surrogate first.
+    {-# INLINE step #-}
+    step UTF16CharDecodeInit x
+         -- XXX Combine the cases and check perf
+         | x < 0xD800 = pure $ ParserD.Done 0 $ unsafeChr $ fromIntegral x
+         | x > 0xDFFF = pure $ ParserD.Done 0 $ unsafeChr $ fromIntegral x
+         | x >= 0xD800 && x <= 0xDBFF = -- high surrogate
+               pure $ ParserD.Partial 0 $ UTF16CharDecoding x
+         | otherwise =
+               let msg = prefix ++ "Invalid first UTF16 word " ++ show x
+                in pure $ handleError msg False
+
+    -- We backtrack here as there is a chance the word in the buffer will
+    -- combine with the word after the current word.
+    step (UTF16CharDecoding prev) x
+         | x >= 0xDC00 && x <= 0xDFFF = -- low surrogate
+               pure $ ParserD.Done 0 $ combineSurrogates prev x
+         | otherwise =
+               let msg = prefix
+                       ++ "Invalid subsequent UTF16 word " ++ show x
+                       ++ " in state " ++ show prev
+                in pure $ handleError msg True
+
+    {-# INLINE extract #-}
+    extract UTF16CharDecodeInit =  error $ prefix ++ "Not enough input"
```

Comment:
We should use Error rather than error, so that the parser can handle the case.

Reply:
This applies to UTF8 as well then